usecases/modules/compare.go

```diff
@@ -0,0 +1,137 @@
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package modules
+
+import (
+	"context"
+
+	"github.com/weaviate/weaviate/entities/additional"
+	"github.com/weaviate/weaviate/entities/models"
+	"github.com/weaviate/weaviate/entities/modulecapabilities"
+	"github.com/weaviate/weaviate/entities/moduletools"
+	"github.com/weaviate/weaviate/entities/schema"
+	"github.com/weaviate/weaviate/entities/search"
+)
+
+func reVectorize(ctx context.Context, cfg moduletools.ClassConfig, mod modulecapabilities.Vectorizer, object *models.Object, class *models.Class, sourceProperties []string, targetVector string, findObjectFn modulecapabilities.FindObjectFn) (bool, models.AdditionalProperties, []float32) {
+	textProps, mediaProps, err := mod.VectorizedProperties(cfg)
+	if err != nil {
+		return true, nil, nil
+	}
+
+	type compareProps struct {
+		Name    string
+		IsArray bool
+	}
+	propsToCmpare := make([]compareProps, 0)
+
+	for _, prop := range class.Properties {
+		if len(prop.DataType) > 1 {
+			continue // multi cref
+		}
+
+		// for named vectors with explicit source properties, skip if not in the list
+		if sourceProperties != nil {
+			found := false
+			for _, sourceProp := range sourceProperties {
+				if prop.Name == sourceProp {
+					found = true
+					break
+				}
+			}
+			if !found {
+				continue
+			}
+		}
+
+		if prop.ModuleConfig != nil {
+			if modConfig, ok := prop.ModuleConfig.(map[string]interface{})[class.Vectorizer]; ok {
+				if skip, ok2 := modConfig.(map[string]interface{})["skip"]; ok2 && skip == true {
+					continue
+				}
+			}
+		}
+
+		if (prop.DataType[0] == schema.DataTypeText.String() || prop.DataType[0] == schema.DataTypeTextArray.String()) && textProps {
+			propsToCmpare = append(propsToCmpare, compareProps{Name: prop.Name, IsArray: schema.IsArrayDataType(prop.DataType)})
+			continue
+		}
+
+		for _, mediaProp := range mediaProps {
+			if mediaProp == prop.Name {
+				propsToCmpare = append(propsToCmpare, compareProps{Name: prop.Name, IsArray: schema.IsArrayDataType(prop.DataType)})
+			}
+		}
+	}
+
+	// if no properties to compare, we can skip the comparison. Return vectors of old object if present
+	if len(propsToCmpare) == 0 {
+		oldObject, err := findObjectFn(ctx, class.Class, object.ID, nil, additional.Properties{}, object.Tenant)
+		if err != nil || oldObject == nil {
+			return true, nil, nil
+		}
+		if targetVector == "" {
+			return false, oldObject.AdditionalProperties, oldObject.Vector
+		} else {
+			return false, oldObject.AdditionalProperties, oldObject.Vectors[targetVector]
+		}
+	}
+
+	returnProps := make(search.SelectProperties, 0, len(propsToCmpare))
+	for _, prop := range propsToCmpare {
+		returnProps = append(returnProps, search.SelectProperty{Name: prop.Name, IsPrimitive: true, IsObject: false})
+	}
+	oldObject, err := findObjectFn(ctx, class.Class, object.ID, returnProps, additional.Properties{}, object.Tenant)
```

Comment:
ideas for future optimizations:
- skip search for `add` object action - old object does not exist
- search only once and reuse object for all target vectors - could be important for `batch add`

Reply:
yeah, using the exists function to make a cheaper existance check would make sense 🤔 

If it is ok, I'd look at these things in the batch-vectorization PR

Reply:
sure, it can be done iteratively.

(what I meant in 2nd point is that each target(named) vector is handled separately in terms of vectorization, so for object of class with 10 vectors configured, previous object would be fetched 10 times, while it could be fetched once and reused for comparison for each of 10 vectors. It's not that this PR changes behaviour, it was already like that originally, but might become performance problem in the future)